client/src/i18n.js

```diff
@@ -169,6 +169,30 @@ class i18n {
       return (map[key]) ? map[key] : match;
     });
   }
+  
+  /**
+   * Substitutes variables with JSX from the given list of injections. This returns an array as JSX
+   * elements cannot be interpolated into strings. React will render arrays like a concatenated
+   * string. You may also use this function like `inject` or you can call `inject` before
+   * `injectJSX` for non-JSX injections.
+   *
+   * @param {String} s - The string to perform the substitutions on.
+   * @param {Object} map - An object with JSX elements to substitute e.g. (var: <b>value</b>}.
+   * @return {Array<JSX.Element|String>}
+   */
+  injectJSX(s, map) {
+    // Split the string by the regex for matches
+    const parts = s.split(this.matchRegex);
+
+    // The capturing group is put in where matches were found - this means all matches are at odd
+    // indexes in the "parts" array. We can replace those items (or leave the "{match}" syntax.
+    for (let i = 1, length = parts.length; i < length; i += 2) {
+      parts[i] = map[parts[i]] || `{${parts[i]}}`;
+    }
```

Comment:
Can we do this in an immutable way? I feel like this is screaming to be a `reduce`.

Reply:
It doesn't need to be immutable because the array is generated above. I don't think we really need that complexity?

Reply:
I guess it's a matter of preference, but I think aiming to use functional programming whenever possible is a good thing.

Reply:
I considered the reduce tbh. The awkwardness was that I only care about every `2n+1`th record though, there's not really any easy way to reduce like that. I can alter it if you feel strongly - I'm not too fussed.

Reply:
All good. 👍